File: cpp/day8/day8.cpp
#include <cstdio>
#include <iostream>
#include <vector>
#include <stack>
#include <map>
#include<fstream>
#include<string>
#include<algorithm>

char* wood = nullptr;
char* vis_grid = nullptr;
int num_cols = 0, num_rows = 0;

std::vector<std::string> read_input_file(const char* file_name) {
    std::vector<std::string> result;
    std::ifstream infile(file_name, std::ifstream::in);
    std::string line;
    while (getline(infile, line)) {
        result.push_back(line);
    }
    infile.close();
    return result;
}

int offset(int col, int row) {
    return row * num_cols + col;
}

void set(int col, int row, char value) {
    wood[offset(col, row)] = value;
}

char get(int col, int row) {
    return wood[offset(col, row)];
}

bool visible(int col, int row) {
    return (vis_grid[offset(col, row)] == 't');
}

void visible(int col, int row, bool value) {
    vis_grid[offset(col, row)] = (value ? 't' : 'f');
}

void setup_wood(const std::vector<std::string>& lines) {
    num_rows = lines.size();
    num_cols = lines[0].size();
    if (wood != nullptr) {
        delete[] wood;
        delete[] vis_grid;
        wood = nullptr;
        vis_grid = nullptr;
    }

    int num_chars = num_cols * num_rows * sizeof(char);
    wood = new char[num_chars];
    vis_grid = new char[num_chars];
    memset(wood, 0, num_chars);
    memset(vis_grid, 'f', num_chars);

    int r = 0, c = 0;
    for (auto line : lines) {
        c = 0;
        for (auto ch : line) {
            set(c++, r, ch - '0');
        }
        r++;
    }
}

void sweep_east() {
    int g, h;
    for (int r = 0; r < num_rows; r++) {
        g = -1;
        for (int c = 0; c < num_cols; c++) {
            h = get(c, r);
            if (h > g) {
                visible(c, r, true);
                g = h;
            }
        }
    }
}

void sweep_west() {
    int g, h;
    for (int r = 0; r < num_rows; r++) {
        g = -1;
        for (int c = num_cols - 1; c >= 0; c--) {
            h = get(c, r);
            if (h > g) {
                visible(c, r, true);
                g = h;
            }
        }
    }
}

void sweep_south() {
    int g, h;
    for (int c = 0; c < num_cols; c++) {
        g = -1;
        for (int r = 0; r < num_rows; r++) {
            h = get(c, r);
            if (h > g) {
                visible(c, r, true);
                g = h;
            }
        }
    }
}

void sweep_north() {
    int g, h;
    for (int c = 0; c < num_cols; c++) {
        g = -1;
        for (int r = num_rows - 1; r >= 0; r--) {
            h = get(c, r);
            if (h > g) {
                visible(c, r, true);
                g = h;
            }
        }
    }
}

void do_sweeps()
{
    sweep_east();
    sweep_north();
    sweep_west();
    sweep_south();
}

int count_visibles()
{
    int count = 0;
    for (int r = 0; r < num_rows; r++) {
        for (int c = 0; c < num_cols; c++) {
            if (visible(c, r)) {
                count++;
            }
        }
    }
    return count;
}

void display()
{
    std::cout << std::endl;
    for (int r = 0; r < num_rows; r++) {
        for (int c = 0; c < num_cols; c++) {
            std::cout << (char)(get(c, r) + '0') << (visible(c, r) ? '+' : '.');
        }
        std::cout << std::endl;
    }
}

int viewing_distance(int col, int row, int direction) {
    int count = 0;
    int h = get(col, row);
    switch (direction) {
    case 0:
        // North
        for (int i = row - 1; i >= 0; i--) {
            count++;
            if (get(col, i) >= h) {
                break;
            }
        }
        break;
    case 1:
        // East
        for (int i = col + 1; i < num_cols; i++) {
            count++;
            if (get(i, row) >= h) {
                break;
            }
        }
        break;
    case 2:
        // South
        for (int i = row + 1; i < num_rows; i++) {
            count++;
            if (get(col, i) >= h) {
                break;
            }
        }
        break;
    case 3:
        // West
        for (int i = col - 1; i >= 0; i--) {
            count++;
            if (get(i, row) >= h) {
                break;
            }
        }
        break;
    }
    return count;
}

int scenic_score(int col, int row) {
    int score = 1;
    for (int direction = 0; direction < 4; direction++) {
        score *= viewing_distance(col, row, direction);
    }
    return score;
}

int best_scenic_score() {
    int best = 0, score = 0;
    for (int row = 0; row < num_rows; row++) {
        for (int col = 0; col < num_cols; col++) {
            score = scenic_score(col, row);
            if (score > best) {
                best = score;
            }
        }
    }
    return best;
}

int main()
{
    std::cout << "Advent of Code 2022\nDay 8 -  Part 1." << std::endl;
    auto lines = read_input_file("day8.txt");
    setup_wood(lines);
    display();
    do_sweeps();
    display();
    int count = count_visibles();
    std::cout << "Result: " << count << " visible sites." << std::endl;

    std::cout << "Part 2." << std::endl;
    int score = best_scenic_score();
    std::cout << "Best scenic score in wood = " << score << std::endl;
}

File: py/day8/day8.py
#
# Day 8 - Treetop Tree House.
# 

import fileinput
from glob import glob
from xmlrpc.client import MAXINT
import numpy as np
from functools import reduce


def read_input_lines(filename):
    lines = []
    with fileinput.FileInput(files = (filename), mode = 'r') as input:
        for raw_line in input:
            line = raw_line.strip(' \n')
            lines.append(line)
    return lines

def prepare_input(lines):
    nr = len(lines)
    nc = len(lines[0])
    elevation = np.zeros((nr, nc), dtype='i')
    for r in range(nr):
        line = lines[r]
        for c in range(nc):
            ch = line[c]
            elevation[r, c] = int(ch)
    return elevation

def visible_north(e, r, c):
    h = e[r, c]
    return (r == 0) or all([e[x, c] < h for x in range(r)])

def visible_east(e, r, c):
    h = e[r, c]
    nc = e.shape[1]
    return (c == nc - 1) or all([e[r, x] < h for x in range(c + 1, nc)])

def visible_south(e, r, c):
    h = e[r, c]
    nr = e.shape[0]
    return (r == nr - 1) or all([e[x, c] < h for x in range(r + 1, nr)])

def visible_west(e, r, c):
    h = e[r, c]
    return (c == 0) or all([e[r, x] < h for x in range(c)])

# On elevation map e, is cell (r, c) visible from outside the grid?
def visible(e, r, c):
    return visible_north(e, r, c) or visible_east(e, r, c) or visible_south(e, r, c) or visible_west(e, r, c)

def visible_tuple(e, item):
    r, c = item
    return visible(e, r, c)

# Find viewing distance on map e, from (r, c) in a given direction dir.
def view_distance(e, r, c, dir):
    count = 0
    h = e[r, c]
    if dir == 0:        # North
        for y in range(r - 1, -1, -1):
            count += 1
            if e[y, c] >= h:
                break
    elif dir == 1:      # East
        for x in range(c + 1, e.shape[1], 1):
            count += 1
            if e[r, x] >= h:
                break
    elif dir == 2:      # South
        for y in range(r + 1, e.shape[0], 1):
            count += 1
            if e[y, c] >= h:
                break
    elif dir == 3:      # West
        for x in range(c - 1, -1, -1):
            count += 1
            if e[r, x] >= h:
                break
    else:
        print('Error in view_distance, dir = ', dir)
    return count

def scenic_score(e, r, c):
    distances = [view_distance(e, r, c, d) for d in [0, 1, 2, 3]]
    return reduce((lambda x, y: x * y), distances, 1)

def scenic_score_tuple(e, item):
    r,c = item
    return scenic_score(e, r, c)

def part1(filename):
    lines = read_input_lines(filename)
    elevation = prepare_input(lines)
    return [visible_tuple(elevation, idx) for idx,_ in np.ndenumerate(elevation) ].count(True)

def part2(filename):
    lines = read_input_lines(filename)
    elevation = prepare_input(lines)
    scores = [scenic_score_tuple(elevation, idx) for idx,_ in np.ndenumerate(elevation) ]
    return max(scores)
   

print('Advent of Code 2022 - Day 8.')
print('Part 1.')
count = part1('day8.txt')
print('Number of visible trees = ', count)

print('Part 2.')
score = part2('day8.txt')
print('Max scenic score for trees in the grid = ', score)
